pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationProvider.java

```diff
@@ -71,6 +71,18 @@
     CompletableFuture<Boolean> canConsumeAsync(TopicName topicName, String role,
             AuthenticationDataSource authenticationData, String subscription);
 
+    /**
+     * Returns authorized roles that can access admin-api for given subscription
+     * 
+     * @param topicName
+     *            the fully qualified topic name associated with the topic.
+     * @param subscription
+     *            the subscription name defined by the client
+     * @return
+     */
+    CompletableFuture<Set<String>> getAuthorizedRolesOnSubscription(TopicName topicName,
```

Comment:
Why getting a list of roles here instead of behaving like the rest of the methods? 

I actually think that this should be handled by `canConsumeAsync()`. Since we might not be able to fetch that info from the Authorization plugin implementation.

Reply:
Actually, broker stores authorization metadata by calling AuthorizationServiceProvider::grantPermission(). and right now, broker is having default ZK-AuthorizationProvider which stores metadata into same global-ZK under policies so, if we want to retrieve then we can see it in policies.

However, if one implements different provider which stores this metadata somewhere else then broker will not have the way to get authorization metadata and user will not be able to see stored authorization-roles.

So, I added API by considering that we can utilize this API later to fetch auth metadata from Authorization provider. As this interface can be implemented by user, ideally we should not try to break so, I tried to keep it generic enough (though I am breaking the interface by adding new API but it seems it might be fine right now). 

Does it make sense? or we still want to change api to `isAuthorizedOnSubscription(..)` instead `getAuthorizedRolesOnSubscription(..)`?

Reply:
I would say to not add a new method, but rather handle that internally in the `canConsume()`. That method is already taking the subscription name (along with the client credentials) and should be able to make the assessment. 

There might be cases in which the permission might be already encrypted into the credentials, or that the check needs to be performed against external system. 

If authorization is stored in default zk-store, then doing `pulsar-admin namespaces policies` will already print the authorized roles.

Reply:
> I would say to not add a new method, but rather handle that internally in the canConsume().

hmm.. yes, we can do that. let me fix it.

> If authorization is stored in default zk-store, then doing pulsar-admin namespaces policies will already print the authorized roles

But what if authorization is not using default zk-store. in that case, we may have to add get() method in auth-provider. but that we can do later as well.

Reply:
> I would say to not add a new method, but rather handle that internally in the canConsume().

@merlimat this will not work because we want to explicitly give access to certain role that can use subscription-based admin-api (eg: reset-cursor, skip) for a given subscription. If the role is not configured/present into authorization list then admin-api request should fail.

whereas `canConsume()` can not fail if `subscription-authorization` list is not configured. If we enforce `subscription-authorization` in `canConsume` method then all existing consumer will start failing because we don't set this permission explicitly.

I can make change in `canConsume()` where if `sub-authorization` is configured then validate role else ignore the check. but we can't use `canConsume()` for admin-api validation (once we have this check then we deprecate `sub-prefix-auth`).

Reply:
>  If we enforce subscription-authorization in canConsume method then all existing consumer will start failing because we don't set this permission explicitly. 

Wouldn't that be only addictive? We have multiple levels in the authorization: 
 * Namespace-wide (`produce`/`consume`)
 * Per-Topic (`produce`/`consume`)
 * Per-Subscription (across topics) (`consume`)

As long as a "principal" is granted permission on any level, it should be allowed to consume messages and perform "consumer-related" admin operations.

That shouldn't break any existing consumer, right?

Reply:
I don't see admin api permissions as different from "consume" permissions. 

If a use has consume access to some topic (on all the subscriptions) than it can "mess up" some other consumer by attaching to the wrong subscription and stealing someone else's messages. 

It's exactly the same as opening the admin API to all the users who have consume access on the topic. 

To solve that, you can : 
 1. Grant per-subscription authorization to each user on its subscription
 2. Remove the per-namespace consume authorization

Reply:
> If a use has consume access to some topic (on all the subscriptions) than it can "mess up" some other consumer by attaching to the wrong subscription and stealing someone else's messages.

which is already incorrect. right? we should have a way to prevent this and with this PR it can be possible. But by default we can't enforce Per-Sub for consumers as it can create issue for existing topics. But may be pulsar can have a flag to enforce it in the system.

> I don't see admin api permissions as different from "consume" permissions.

Actually then we don't need this PR at all. No need of introducing "Per-Subscription" authorization if "Namespace-auth(consume)" is anyway present so, "Namespace-auth" will always win and all subscribers can access each other's cursor state. 
So, I thought admin-api right now doesn't have any authorization present, we can make it right atleast for admin-api by enforcing "Per-Subscription" validation. 

But Let me know if you still think admin-api should not enforce "Per-subscription" then I will make your suggested change and all api will use one `canConsume()-api`.

Reply:
> which is already incorrect. right? we should have a way to prevent this and with this PR it can be possible. 

I think having per-subscription ACLs should be enough, right? 

> But by default we can't enforce Per-Sub for consumers as it can create issue for existing topics. But may be pulsar can have a flag to enforce it in the system.

Even for existing topic, we can just add the per-subscription ACLs, no? 

> Actually then we don't need this PR at all. No need of introducing "Per-Subscription" authorization if "Namespace-auth(consume)" is anyway present so, "Namespace-auth" will always win and all subscribers can access each other's cursor state.

But you just need to grant one permission, either at the namespace level or at the subscription level if you want to have a user to be restricted. 

I think this change is good. Otherwise having 2 different ways to enforce authorization for consume vs admin makes things very confusing.

Let's stick with 3 levels of authorization (namespace, topic, subscription). Whoever has "consume" access on any of these, will be able to consume and perform admin related operations on that topic/subscription.

Reply:
>  Whoever has "consume" access on any of these, will be able to consume and perform admin related operations on that topic/subscription.

Alright. I will make the change.

Reply:
@merlimat made the change.